src/elastic-search/query-builder.service.ts

```diff
@@ -0,0 +1,98 @@
+import { Injectable, Logger } from "@nestjs/common";
+import { QueryDslQueryContainer } from "@elastic/elasticsearch/lib/api/types";
+import { IDatasetFields } from "src/datasets/interfaces/dataset-filters.interface";
+
+interface Filter {
+  terms?: {
+    [key: string]: string[] | boolean[];
+  };
+  term?: {
+    [key: string]: string | boolean;
+  };
+}
+
+const addTermsFilter = (
+  fieldName: string,
+  values: unknown,
+  filterArray: Filter[],
+) => {
+  if (Array.isArray(values) && values.length > 0) {
+    filterArray.push({
+      terms: {
+        [fieldName]: values,
+      },
+    });
+  }
+  if (typeof values === "boolean") {
+    filterArray.push({
+      term: {
+        [fieldName]: values,
+      },
+    });
+  }
+};
+@Injectable()
+export class SearchQueryBuilderService {
+  public buildSearchQuery(searchParam: IDatasetFields) {
+    const { text = "", ...fields } = searchParam;
+    const filterFields = [
+      "keywords",
+      "type",
+      "creationLocation",
+      "ownerGroup",
+      "isPublished",
+    ];
+    const queryFields = ["datasetName", "description"];
+
+    try {
+      const query: QueryDslQueryContainer[] = [];
+      const filter: Filter[] = [];
+
+      for (const fieldName of filterFields) {
+        addTermsFilter(fieldName, fields[fieldName], filter);
+      }
+
+      if (!text) {
+        return {
+          query: {
+            bool: {
+              filter: filter,
+            },
+          },
+        };
+      }
+
+      const searchTermArray = text.toLowerCase().split(" ");
+      const wildcardQueries = searchTermArray.flatMap((term) =>
+        queryFields.map((fieldName) => ({
+          wildcard: {
+            [fieldName]: {
+              value: `*${term}*`,
+            },
+          },
+        })),
+      );
```

Comment:
Can we submit the query as it is to elastic search without splitting it in terms?

Reply:
We can do that, but then we lose the benefit of partial match with multiple terms and list results based on relevancy score.
One solution would be we provide a toggle button on the filter UI and let the user to decide whether they want to splitting terms or not.

Reply:
Either way is fine by me, but if we split the query, we have to do something more elaborate according to NLP methodologies!!!
Have you checked if ES can perform the query splitting internally?